cmd/config.go

```diff
@@ -124,14 +136,95 @@ type HostnamePolicyConfig struct {
+	CertFile string `validate:"required"`
+	KeyFile  string `validate:"required"`
+	// The CACertFile file may contain any number of root certificates and will
+	// be deduplicated internally.
 	CACertFile string `validate:"required"`
+	// Valid ClientAuth values can be found at
+	// https://pkg.go.dev/crypto/tls#ClientAuthType. Boulder will select its own
+	// default value, rather than the default value from //crypto/tls, if no
+	// client auth is configured
+	ClientAuth string `validate:"omitempty"`
+	// Each CipherSuites value must be supported by //crypto/tls. Boulder will
+	// select its own default value, rather than the default value from
+	// //crypto/tls, if no cipher suite(s) is configured. Cipher suites will be
+	// ignored by tls.Config for TLS v1.3. Valid cipher suites can be found at
+	// https://cs.opensource.google/go/go/+/refs/tags/go1.22.2:src/crypto/tls/cipher_suites.go;l=59-68
+	CipherSuites []string `validate:"omitempty"`
+}
+
+// makeCipherSuitesFromConfig takes a slice of human-readable TLS cipher suite
+// names declared in a config file and constructs a new slice of cipher suites
+// usable by a tls.Config or returns an error.
+func (t *TLSConfig) makeCipherSuitesFromConfig() ([]uint16, error) {
+	// We'll set a sane default, rather than use the default from //crypto/tls.
+	if len(t.CipherSuites) == 0 {
+		return []uint16{tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256}, nil
+	}
+
+	// Construct a reverse lookup to validate the operator provided name(s) is
+	// usable. The "legacy names" are unfortunately not included.
+	cipherSuiteNameToId := make(map[string]uint16, len(tls.CipherSuites()))
+	for _, id := range tls.CipherSuites() {
+		cipherSuiteNameToId[id.Name] = id.ID
+	}
+
+	var configuredCipherSuiteIds []uint16
+	for _, cs := range t.CipherSuites {
+		id, ok := cipherSuiteNameToId[cs]
+		if !ok {
+			return nil, fmt.Errorf("unsupported TLS cipher suite: %s", cs)
+		}
+		configuredCipherSuiteIds = append(configuredCipherSuiteIds, id)
+	}
+
+	return configuredCipherSuiteIds, nil
+}
+
+// makeClientAuthFromConfig converts the value of clientAuth from config file
+// and returns the equivalent tls.ClientAuthType value or an error.
+func (t *TLSConfig) makeClientAuthFromConfig() (tls.ClientAuthType, error) {
+	// Construct some lookup tables to map the configured ClientAuthType to the
+	// appropriate //crypto/tls const integer.
+	clientAuthTypeNameToId := make(map[string]int)
+	clientAuthTypeIdToName := make(map[int]string)
+
+	for i := 0; ; i++ {
+		name := tls.ClientAuthType(i).String()
+		if strings.Contains(name, "ClientAuthType") {
+			// A const block using iota doesn't have length so to speak.
+			// Non-existent values will return the name of the underlying type
+			// (not the concrete type!). We can use that to bail out early as
+			// soon as our maps contain only the values used inside
+			// //crypto/tls.
+			break
+		}
+		clientAuthTypeNameToId[name] = i
+		clientAuthTypeIdToName[i] = name
+	}
+
```

Comment:
Phil and I talked offline about the added support for configuring cipher suites. This was added for a downstream proxy which did not support our selected TLS v1.2 cipher-suite. There are two alternatives to this approach:

1. Pick a cipher-suite and construct a modified TLSConfig in `remoteva` for just this purpose.
2. Update our grpc-go credentials code to only use TLS v1.3, it's currently set to v1.2. This version [was selected 8 years ago](https://github.com/letsencrypt/boulder/pull/2307/files#diff-48f4f49996ea7be2c77a2217b8d0d4584f312961d1dc3d952e57354c85563d05R84), which was 2 years before RFC 8446 was published, August 2018.